### PR TITLE
Discussion about further improvements for edge and node classes

### PR DESCRIPTION
**Pull request by sven5s**
edge.classes is now a string containing multiple classes, so looking for substring instead of exact match, may switch to list or tuple in future

### CODE REVIEW

File: ipycytoscape/cytoscape.py [hunk 1]
```diff
@@ -381,7 +381,7 @@ def remove_edge(self, edge):
         try:
             self.edges.remove(edge)
             del self._adj[edge.data['source']][edge.data['target']]
-            if not edge.classes == 'directed':
+            if not 'directed' in edge.classes :
```
Comment (ianhi) [inline ipycytoscape/cytoscape.py:384]:
```suggestion
            if not 'directed' in edge.classes:
```

**Review by ianhi — COMMENTED** [review]
This is a clear improvement thanks! One small thing to change and then I think this is good. 
Two potential followups:

1. Do you have a small example demonstrating the issue, I'd like to add it as a test
2. A related issue is making the class list into a list rather than a string, see https://github.com/QuantStack/ipycytoscape/issues/100

File: ipycytoscape/cytoscape.py (hunk 1 above)
Comment (sven5s) [reply ipycytoscape/cytoscape.py:384]: I can provide an example tomorrow. But I discovered this issue rather simply. Import a graph with directed=True for example from json (maybe it's easier to create something arbitary with networkx) and try do delete an edge.

Mh, if I have time, I will try do convert this string to an list. This should be straightforward.

**Comment by sven5s** [comment]
I can provide an example tomorrow. But I discovered this issue rather simply. Import a graph with directed=True for example from json (maybe it's easier to create something arbitary with networkx) and try do delete an edge.

Mh, if I have time, I will try do convert this string to an list. This should be straightforward.

**Comment by sven5s** [comment]
Small example to test the issue:
```
import ipycytoscape
data = {
    'nodes': [
        { 'data': { 'id': 'a'} },
        { 'data': { 'id': 'b'} }
    ],
    'edges': [
        {'data': { 'source': 'a', 'target': 'b' } }
    ]
}

graph = ipycytoscape.Graph()
graph.add_graph_from_json(data, directed=True)

graph.remove_edge(graph.edges[0])
```

**Comment by marimeireles** [comment]
Thanks for your changes @sven5s!
The problem here is that our tests were expecting strings and not lists and this change broke a lot of them. Let me know if you're up for fixing it or if we should do it as a joint effort somehow :)
Also, it'd be cool to include this(https://github.com/QuantStack/ipycytoscape/pull/162#issuecomment-698145236) as a test with a comment or a self-explanatory name of what this test is doing I think.
Thanks again! :cherry_blossom:

**Comment by sven5s** [comment]
I think the test should be fixed now.
In my opinion it would be better to test the graph manipulation in general (add nodes and egdes, remove nodes and edges). My Code is only a small example.

**Comment by sven5s** [comment]
Argh, list are not compatible with the actual json format. Should the string be converted or the json just altered?

**Comment by marimeireles** [comment]
I think if we change the json we break compatibility with the "cytoscape json" right @ianhi?
Don't think we should do it.

**Comment by ianhi** [comment]
>  In my opinion it would be better to test the graph manipulation in general (add nodes and egdes, remove nodes and edges). My Code is only a small example. But I'm not very familiar with pytest.

This is a great point. If you're motivated to add more tests (as part of this or another PR) I'm happy to help advise on how to do it. I basically just run something in a notebook and print out the objects in order to generate what the expected outcome is. See description here: https://github.com/QuantStack/ipycytoscape/pull/146#issuecomment-683546916

> Argh, list are not compatible with the actual json format. Should the string be converted or the json just altered?

> I think if we change the json we break compatibility with the "cytoscape json" right @ianhi?

We would also need to change this line:
https://github.com/QuantStack/ipycytoscape/blob/fd8b4bdd58f2e407f64cbe3b76c01afe40dbf390/src/widget.ts#L93

to be `classes: []` and then I think this 

As for compatibility, I think that either strings or lists are compatible with what cytoscape accepts but I'm worried about backwards compatibility. For example, all my code using ipycytoscape will break if we switch from strings to lists for classes. I think there are three options:
1. Support both lists and strings for classes
    - This would be fully compatible with cytoscapejs
2. Switch to only allowing lists for classes and bump to version `2.0.0`
3. Stick to only strings


Option 1 is the most work, but I think it is also the best solution. So I guess my  vote would be: If @sven5s is interested lets implement option 1 here, otherwise lets stick to strings for now and implement option 1 or option 2 later.

**Comment by ianhi** [comment]
Oh also if we change the way classes work edges then the nodes should have the same done to their classes.

**allowing multiple types**

To allow multiple types we will need to use the traitlets union on the python side:
`Union(Unicode(), List())`

Also, scratch what I said earlier about the changes on the typescript side. All that the line `classes: ''` does is set a default value, it can be overwritten no problem.

**Comment by sven5s** [comment]
Argh, this becomes really difficult. We should discuss a little further here, close this pull request and I will make a clean one.
In my opinion, this will become very nasty over time and it will be hard to maintain. It's also not very pythonic. The logic to manipulate edges and nodes should be moved to the corresponding class. The internal representation isn't important anymore an the logic can be move outside of the graph.
But first #109 should be merged. In the mean time i will make a pull request with the necessary bug fixes. Then we can move on to change the edge and node classes.

How sounds this?

P.S.: It will also become easier to create nodes and edges to add to the graph. At the moment it's not very comfortable.

**Comment by sven5s** [comment]
Great #109 was merged. Now is the question if a string or a list should be used for the classes (for backward compatibility a string is better and the difference for this usecase is not that big) and if the logic for the properties should be merged in the Edge and Node classes. These properties would be the id for nodes and source, target, multiple_edges and directed for edges. The classes string and data dict will still persist, only getter and setter functions will be offered.
For example it's then possible to do
```
Node(id=...)
node.id = ...
edge.source = ...
Edge(source=..., target=..., directed=..., multiple_edges=...)
edge.directed = True
if edge.directed:
    ...
```
And also _set_attributes should be moved in Node and Edge classes (maybe Element).

I like the idea and think that the usability would be improved. What's your opinion?

**Comment by ianhi** [comment]
Hi @sven5s sorry for having dropped this for so long. This has been a wild week in lab for me but I'll try to look at this this weekend. If that doesn't happen feel free to ping this issue again. 

Also, this will need a rebase now. (https://medium.com/@dirk.avery/the-definitive-git-rebase-guide-dbd7717f9437)

**Comment by sven5s** [comment]
Hy @ianhi,
I don't think a rebase is useful actual. Please take a look at my comments https://github.com/QuantStack/ipycytoscape/pull/162#issuecomment-698557163 and https://github.com/QuantStack/ipycytoscape/pull/162#issuecomment-717342901 above.

**Comment by sven5s** [comment]
@marimeireles and @ianhi 
I really want to help you with that great project but I don't want to make design decision and implement it if they will not be merged later. In the case it would be much easier for me to write some simple function around the existing API. But I want to help you to improve it.

**Comment by marimeireles** [comment]
Thanks @sven5s ! Sorry for leaving this open for so long. I'm kinda lost though on your other PR. Commented there :)

**Pull request closed without merging**